// @julep/sdk/resources/executions/executions.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import { isRequestOptions } from "../../core.mjs";
import * as TransitionsAPI from "./transitions.mjs";
import { Transitions, } from "./transitions.mjs";
import { OffsetPagination } from "../../pagination.mjs";
export class Executions extends APIResource {
    constructor() {
        super(...arguments);
        this.transitions = new TransitionsAPI.Transitions(this._client);
    }
    /**
     * Create Task Execution
     */
    create(taskId, body, options) {
        return this._client.post(`/tasks/${taskId}/executions`, { body, ...options });
    }
    list(taskId, query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list(taskId, {}, query);
        }
        return this._client.getAPIList(`/tasks/${taskId}/executions`, ExecutionsOffsetPagination, {
            query,
            ...options,
        });
    }
    /**
     * Update Execution
     */
    changeStatus(executionId, body, options) {
        return this._client.put(`/executions/${executionId}`, { body, ...options });
    }
    /**
     * Get Execution Details
     */
    get(executionId, options) {
        return this._client.get(`/executions/${executionId}`, options);
    }
}
export class ExecutionsOffsetPagination extends OffsetPagination {
}
export class TransitionsOffsetPagination extends OffsetPagination {
}
Executions.ExecutionsOffsetPagination = ExecutionsOffsetPagination;
Executions.Transitions = Transitions;
