// @julep/sdk/resources/executions/executions.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransitionsOffsetPagination = exports.ExecutionsOffsetPagination = exports.Executions = void 0;
const resource_1 = require("../../resource.js");
const core_1 = require("../../core.js");
const TransitionsAPI = __importStar(require("./transitions.js"));
const transitions_1 = require("./transitions.js");
const pagination_1 = require("../../pagination.js");
class Executions extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.transitions = new TransitionsAPI.Transitions(this._client);
    }
    /**
     * Create Task Execution
     */
    create(taskId, body, options) {
        return this._client.post(`/tasks/${taskId}/executions`, { body, ...options });
    }
    list(taskId, query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list(taskId, {}, query);
        }
        return this._client.getAPIList(`/tasks/${taskId}/executions`, ExecutionsOffsetPagination, {
            query,
            ...options,
        });
    }
    /**
     * Update Execution
     */
    changeStatus(executionId, body, options) {
        return this._client.put(`/executions/${executionId}`, { body, ...options });
    }
    /**
     * Get Execution Details
     */
    get(executionId, options) {
        return this._client.get(`/executions/${executionId}`, options);
    }
}
exports.Executions = Executions;
class ExecutionsOffsetPagination extends pagination_1.OffsetPagination {
}
exports.ExecutionsOffsetPagination = ExecutionsOffsetPagination;
class TransitionsOffsetPagination extends pagination_1.OffsetPagination {
}
exports.TransitionsOffsetPagination = TransitionsOffsetPagination;
Executions.ExecutionsOffsetPagination = ExecutionsOffsetPagination;
Executions.Transitions = transitions_1.Transitions;
