// @julep/sdk/resources/sessions.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionsOffsetPagination = exports.Sessions = void 0;
const resource_1 = require("../resource.js");
const core_1 = require("../core.js");
const pagination_1 = require("../pagination.js");
class Sessions extends resource_1.APIResource {
    /**
     * Create Session
     */
    create(body, options) {
        return this._client.post('/sessions', { body, ...options });
    }
    /**
     * Patch Session
     */
    update(sessionId, body, options) {
        return this._client.patch(`/sessions/${sessionId}`, { body, ...options });
    }
    list(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/sessions', SessionsOffsetPagination, { query, ...options });
    }
    /**
     * Delete Session
     */
    delete(sessionId, options) {
        return this._client.delete(`/sessions/${sessionId}`, options);
    }
    /**
     * Initiates a chat session.
     *
     * Parameters: developer (Developer): The developer associated with the chat
     * session. session_id (UUID): The unique identifier of the chat session.
     * chat_input (ChatInput): The chat input data. background_tasks (BackgroundTasks):
     * The background tasks to run. x_custom_api_key (Optional[str]): The custom API
     * key.
     *
     * Returns: ChatResponse: The chat response.
     */
    chat(sessionId, params, options) {
        const { connection_pool, 'X-Custom-Api-Key': xCustomAPIKey, ...body } = params;
        return this._client.post(`/sessions/${sessionId}/chat`, {
            query: { connection_pool },
            body,
            ...options,
            headers: {
                ...(xCustomAPIKey != null ? { 'X-Custom-Api-Key': xCustomAPIKey } : undefined),
                ...options?.headers,
            },
        });
    }
    /**
     * Create Or Update Session
     */
    createOrUpdate(sessionId, body, options) {
        return this._client.post(`/sessions/${sessionId}`, { body, ...options });
    }
    /**
     * Get Session
     */
    get(sessionId, options) {
        return this._client.get(`/sessions/${sessionId}`, options);
    }
    /**
     * Get Session History
     */
    history(sessionId, options) {
        return this._client.get(`/sessions/${sessionId}/history`, options);
    }
    /**
     * Renders a chat input.
     *
     * Parameters: developer (Developer): The developer associated with the chat
     * session. session_id (UUID): The unique identifier of the chat session.
     * chat_input (ChatInput): The chat input data.
     *
     * Returns: RenderResponse: The rendered chat input.
     */
    render(sessionId, body, options) {
        return this._client.post(`/sessions/${sessionId}/render`, { body, ...options });
    }
    /**
     * Update Session
     */
    reset(sessionId, body, options) {
        return this._client.put(`/sessions/${sessionId}`, { body, ...options });
    }
}
exports.Sessions = Sessions;
class SessionsOffsetPagination extends pagination_1.OffsetPagination {
}
exports.SessionsOffsetPagination = SessionsOffsetPagination;
Sessions.SessionsOffsetPagination = SessionsOffsetPagination;
