// @julep/sdk/resources/users/docs.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocsOffsetPagination = exports.Docs = void 0;
const resource_1 = require("../../resource.js");
const core_1 = require("../../core.js");
const docs_1 = require("../docs.js");
Object.defineProperty(exports, "DocsOffsetPagination", { enumerable: true, get: function () { return docs_1.DocsOffsetPagination; } });
class Docs extends resource_1.APIResource {
    /**
     * Creates a new document for a user.
     *
     * Parameters: user_id (UUID): The unique identifier of the user associated with
     * the document. data (CreateDocRequest): The data to create the document with.
     * x_developer_id (UUID): The unique identifier of the developer associated with
     * the document.
     *
     * Returns: Doc: The created document.
     */
    create(userId, params, options) {
        const { connection_pool, ...body } = params;
        return this._client.post(`/users/${userId}/docs`, { query: { connection_pool }, body, ...options });
    }
    list(userId, query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list(userId, {}, query);
        }
        return this._client.getAPIList(`/users/${userId}/docs`, docs_1.DocsOffsetPagination, { query, ...options });
    }
    /**
     * Delete User Doc
     */
    delete(userId, docId, options) {
        return this._client.delete(`/users/${userId}/docs/${docId}`, options);
    }
    /**
     * Searches for documents associated with a specific user.
     *
     * Parameters: x_developer_id (UUID): The unique identifier of the developer
     * associated with the user. search_params (TextOnlyDocSearchRequest |
     * VectorDocSearchRequest | HybridDocSearchRequest): The parameters for the search.
     * user_id (UUID): The unique identifier of the user associated with the documents.
     *
     * Returns: DocSearchResponse: The search results.
     */
    search(userId, params, options) {
        const { connection_pool, ...body } = params;
        return this._client.post(`/users/${userId}/search`, { query: { connection_pool }, body, ...options });
    }
}
exports.Docs = Docs;
