// @julep/sdk/resources/users/docs.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import { isRequestOptions } from "../../core.mjs";
import { DocsOffsetPagination } from "../docs.mjs";
export class Docs extends APIResource {
    /**
     * Creates a new document for a user.
     *
     * Parameters: user_id (UUID): The unique identifier of the user associated with
     * the document. data (CreateDocRequest): The data to create the document with.
     * x_developer_id (UUID): The unique identifier of the developer associated with
     * the document.
     *
     * Returns: Doc: The created document.
     */
    create(userId, params, options) {
        const { connection_pool, ...body } = params;
        return this._client.post(`/users/${userId}/docs`, { query: { connection_pool }, body, ...options });
    }
    list(userId, query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list(userId, {}, query);
        }
        return this._client.getAPIList(`/users/${userId}/docs`, DocsOffsetPagination, { query, ...options });
    }
    /**
     * Delete User Doc
     */
    delete(userId, docId, options) {
        return this._client.delete(`/users/${userId}/docs/${docId}`, options);
    }
    /**
     * Searches for documents associated with a specific user.
     *
     * Parameters: x_developer_id (UUID): The unique identifier of the developer
     * associated with the user. search_params (TextOnlyDocSearchRequest |
     * VectorDocSearchRequest | HybridDocSearchRequest): The parameters for the search.
     * user_id (UUID): The unique identifier of the user associated with the documents.
     *
     * Returns: DocSearchResponse: The search results.
     */
    search(userId, params, options) {
        const { connection_pool, ...body } = params;
        return this._client.post(`/users/${userId}/search`, { query: { connection_pool }, body, ...options });
    }
}
export { DocsOffsetPagination };
