// @julep/sdk/resources/users/users.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsersOffsetPagination = exports.Users = void 0;
const resource_1 = require("../../resource.js");
const core_1 = require("../../core.js");
const DocsAPI = __importStar(require("./docs.js"));
const docs_1 = require("./docs.js");
const pagination_1 = require("../../pagination.js");
class Users extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.docs = new DocsAPI.Docs(this._client);
    }
    /**
     * Create User
     */
    create(body, options) {
        return this._client.post('/users', { body, ...options });
    }
    /**
     * Patch User
     */
    update(userId, body, options) {
        return this._client.patch(`/users/${userId}`, { body, ...options });
    }
    list(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/users', UsersOffsetPagination, { query, ...options });
    }
    /**
     * Delete User
     */
    delete(userId, options) {
        return this._client.delete(`/users/${userId}`, options);
    }
    /**
     * Create Or Update User
     */
    createOrUpdate(userId, body, options) {
        return this._client.post(`/users/${userId}`, { body, ...options });
    }
    /**
     * Get User Details
     */
    get(userId, options) {
        return this._client.get(`/users/${userId}`, options);
    }
    /**
     * Update User
     */
    reset(userId, body, options) {
        return this._client.put(`/users/${userId}`, { body, ...options });
    }
}
exports.Users = Users;
class UsersOffsetPagination extends pagination_1.OffsetPagination {
}
exports.UsersOffsetPagination = UsersOffsetPagination;
Users.UsersOffsetPagination = UsersOffsetPagination;
Users.Docs = docs_1.Docs;
